lessons/bn/ecto/changesets.md

```diff
@@ -0,0 +1,291 @@
+  valid?: true
+>
+```
+
+এক্টো বলছে, চেইঞ্জসেটটি ভ্যালিড, কিন্তু আসলে আমরা ফাঁকা নাম অনুমোদন করি না, চলুন এটা ঠিক করা যাক!
+
+## ভ্যালিডেশান
+
+ভ্যালিডেশান এর জন্যে এক্টোতে অনেকগুলো বিল্ট-ইন ফাংশন রয়েছে।
+
+আমরা `Ecto.Changeset` বহুলভাবে ব্যবহার করতে যাচ্ছি, তাই আমরা `Ecto.Changeset` কে আমাদের `person.ex` মডিউলে ইম্পোর্ট করবো, যেটাতে আমাদের স্কিমা ও রয়েছে:
+
+```elixir
+defmodule Friends.Person do
+  use Ecto.Schema
+  import Ecto.Changeset
+
+  schema "people" do
+    field :name, :string
+    field :age, :integer, default: 0
+  end
+end
+```
+
+এখন, আমরা `cast/3` ফাংশনটি সরাসরি ব্যবহার করতে পারি।
+
+একটি স্কিমার এক বা একাধিক চেইঞ্জসেট ক্রিয়েটর ফাংশন থাকতে পারে। চলুন, একটি ক্রিয়েটর ফাংশন তৈরি করি যেটা একটি স্ট্রাক্ট ও চেইঞ্জের ম্যাপ নেয় এবং চেইঞ্জসেট রিটার্ন করে:
+
+```elixir
+def changeset(struct, params) do
+  struct
+  |> cast(params, [:name, :age])
+end
+```
+
+এখন, চেইঞ্জসেটে সবসময় নাম থাকবে তা আমরা নিশ্চিত করতে পারি:
+
+```elixir
+def changeset(struct, params) do
+  struct
+  |> cast(params, [:name, :age])
+  |> validate_required([:name])
+end
+```
+
+যখন আমরা `Friends.Person.changeset/2` ফাংশনটিকে ফাঁকা নাম দিয়ে কল করি, তখন চেইঞ্জসেট আর ভ্যালিড থাকবে না, তাই এটা একটা এরর মেসেজ দেখিয়ে সাহায্য করবে।
+নোট: `iex` এ কাজ করার সময় `recompile()` রান করতে ভুলবেন না, না হলে কোডের পরিবর্তনগুলো দেখতে পাবেন না।
+
+```elixir
+iex> Friends.Person.changeset(%Friends.Person{}, %{"name" => ""})
+%Ecto.Changeset<
+  action: nil,
+  changes: %{},
+  errors: [name: {"can't be blank", [validation: :required]}],
+  data: %Friends.Person<>,
+  valid?: false
+>
+```
+
+উপরে উল্লেখিত চেইঞ্জসেট দিয়ে যদি আপনি `Repo.insert(changeset)` করতে চান, তাহলে আপনি একই এররটি এবং সাথে `{:error, changeset}` রিসিভ করবেন, তাই আপনাকে প্রতিবারেই `changeset.valid?` ব্যবহার করা লাগবে না।
+সরাসরি, ইনসার্ট, আপডেট ও ডিলিট করার চেষ্টা করে কোনো এরর পেলে সেটা প্রসেস করাই সহজতর।
+
+`validate_required/2` ছাড়াও, `validate_length/3` নামের একটি ফাংশন রয়েছে, যেটা কিছু অতিরিক্ত অপশন নিতে পারে:
+
+```elixir
+def changeset(struct, params) do
+  struct
+  |> cast(params, [:name, :age])
+  |> validate_required([:name])
+  |> validate_length(:name, min: 2)
+end
+```
+
+যদি আমরা একটা মাত্র ক্যারেক্টার বিশিষ্ট নাম প্রদান করি, তাহলে আপনি চেষ্টা করে রেজাল্টটি কি হবে তা অনুমান করে ফেলতে পারেন! 
+
+```elixir
+iex> Friends.Person.changeset(%Friends.Person{}, %{"name" => "A"})
+%Ecto.Changeset<
+  action: nil,
+  changes: %{name: "A"},
+  errors: [
+    name: {"should be at least %{count} character(s)",
+     [count: 2, validation: :length, kind: :min, type: :string]}
+  ],
+  data: %Friends.Person<>,
+  valid?: false
+>
+```
+
+আপনি হয়তো আশ্চর্য হচ্ছেন, এরর মেসেজে বিদঘুটে `%{count}` দেখে — এটা অন্যান্য ল্যাংগুয়েজে ট্রান্সলেশান সহজ করার জন্যেই বিদ্যমান; আপনি যদি সরাসরি ইউজারকে এরর দেখাতে চান, সেক্ষেত্রে আপনি এটাকে [`traverse_errors/2`](https://hexdocs.pm/ecto/Ecto.Changeset.html#traverse_errors/2) ব্যবহার করে মানুষের পড়ার উপযুক্ত করতে তুলতে পারেন — ডকুমেন্টে উল্লেখিত উদাহরণটি দেখে নিন।
```

Comment:
I would suggest a decomposition of the complex sentence like this: "এরর মেসেজে বিদঘুটে `%{count} দেখে হয়ত আপনি আশ্চর্য হচ্ছেন। এটি অন্য়ান্য ভাষায় অনুবাদ সহজ করার জন্য ব্যবহার করা হয়।..."

Reply:
Thanks, I've made slight change in the second sentence: `এরর মেসেজে বিদঘুটে `%{count} দেখে হয়ত আপনি আশ্চর্য হচ্ছেন। এটা অন্যান্য ভাষায় অনুবাদ করা সহজ করে।` 
Let me know if it's okay.

Reply:
@wasi0013 bhai, your version sounds better :D